# "Why So Serious?" - Outage Duration Level Classifier

**Name(s)**: Yi-Hsuan Kuo, Tony Zheng

**Website Link**: https://tzheng846.github.io/Power_Outage_Analysis/

In [304]:
import pandas as pd
import numpy as np
from pathlib import Path
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
pd.options.plotting.backend = 'plotly'


# from dsc80_utils import * # Feel free to uncomment and use this.

In [305]:
import plotly.io as pio
pio.renderers.default = 'jupyterlab'

In [306]:
from datetime import datetime, time

## Step 1: Introduction

(write some introduction)

- We will be answering the following question: 
What are the characteristics of different levels of outage duration?

## Step 2: Data Cleaning and Exploratory Data Analysis

### Data Cleaning

- load the dataset

In [307]:
original = pd.read_csv('data/power_outage.csv')

- select the columns that are relevant to our question

In [308]:
outage = original[['YEAR','MONTH','U.S._STATE','NERC.REGION', 'CLIMATE.REGION', 
                   'ANOMALY.LEVEL', 'CLIMATE.CATEGORY','OUTAGE.START.DATE', 'OUTAGE.START.TIME', 
                   'CAUSE.CATEGORY', 'CAUSE.CATEGORY.DETAIL','HURRICANE.NAMES', 'OUTAGE.DURATION', 
                   'DEMAND.LOSS.MW','CUSTOMERS.AFFECTED', 'RES.PRICE', 'COM.PRICE', 
                   'IND.PRICE','TOTAL.PRICE','TOTAL.CUSTOMERS', 'POPULATION', 'AREAPCT_URBAN', 'PCT_LAND']]

- Cleaning the data

In [309]:
outage.dropna(subset=['OUTAGE.DURATION'], inplace=True)

/var/folders/s9/b35x2lvd0zl66546s25wwsjc0000gn/T/ipykernel_92343/3635334018.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



1. strip 'CAUSE.CATEGORY.DETAIL'
2. Change the zeros in 'DEMAND.LOSS.MW' and 'CUSTOMERS.AFFECTED' to nan because it's unlikely for these values to be 0 which means they should be the missing value. 
3. Change 'OUTAGE.START.DATE' to datetime object and 'OUTAGE.START.TIME' datetime object with only time
4. Create a new column of 'DAY.OR.NIGHT' which tells whether the outage begans on day time or night time.
5. Create a new column of 'HAS.HURRICANE' from 'HURRICANE.NAMES' as a missing value implies there's no hurricane while the outage happened.
6. Drop the 'HURRICANE.NAMES' column as we no longer need its information.

In [310]:
outage['CAUSE.CATEGORY.DETAIL']=outage['CAUSE.CATEGORY.DETAIL'].apply(lambda x: None if pd.isna(x) else x.strip())
outage['DEMAND.LOSS.MW']=outage['DEMAND.LOSS.MW'].apply(lambda x: np.nan if x==0 else x)
outage['CUSTOMERS.AFFECTED']=outage['CUSTOMERS.AFFECTED'].apply(lambda x: np.nan if x==0 else x)
outage['OUTAGE.START.DATE']=outage['OUTAGE.START.DATE'].apply(pd.to_datetime)
outage['OUTAGE.START.TIME']=outage['OUTAGE.START.TIME'].apply(lambda x: datetime.strptime(x,'%I:%M:%S %p').time())
outage['DAY.OR.NIGHT']=outage['OUTAGE.START.TIME'].apply(lambda x: x if pd.isna(x) else('DAY' if x>time(9) and x<time(17) else 'NIGHT'))
outage['HAS.HURRICANE']=outage['HURRICANE.NAMES'].notna()
outage=outage.drop('HURRICANE.NAMES', axis=1)

/var/folders/s9/b35x2lvd0zl66546s25wwsjc0000gn/T/ipykernel_92343/2444993277.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

/var/folders/s9/b35x2lvd0zl66546s25wwsjc0000gn/T/ipykernel_92343/2444993277.py:2: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

/var/folders/s9/b35x2lvd0zl66546s25wwsjc0000gn/T/ipykernel_92343/2444993277.py:3: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the d

- Create a dictionary that classifies durations into 5 categories: 'Momentary', 'Short-term', 'Extended', 'Prolonged', and 'Chronic' base on their length which we believe implies the seriousness of that outage because longer outage usually cause more trouble than shorter ones. 

In [311]:
duration_level={(0, 180): 'Momentary', (181, 720): 'Short-term', (721, 1440): 'Extended', (1441, 4320): 'Prolonged', (4320, np.inf): 'Chronic'}
def duration_classify(duration):
    for range, level in duration_level.items():
        if range[0]<=duration<=range[1]:
            return level
    return 'Other'

In [312]:
outage['duration level']=outage['OUTAGE.DURATION'].apply(duration_classify)

In [355]:
outage.head()

,YEAR,MONTH,U.S._STATE,NERC.REGION,CLIMATE.REGION,ANOMALY.LEVEL,CLIMATE.CATEGORY,OUTAGE.START.DATE,OUTAGE.START.TIME,CAUSE.CATEGORY,...,COM.PRICE,IND.PRICE,TOTAL.PRICE,TOTAL.CUSTOMERS,POPULATION,AREAPCT_URBAN,PCT_LAND,DAY.OR.NIGHT,HAS.HURRICANE,duration level
0,2011,7.0,Minnesota,MRO,East North Central,-0.3,normal,2011-07-01,17:00:00,severe weather,...,9.18,6.81,9.28,2595696,5348119,2.14,91.592666,NIGHT,False,Prolonged
1,2014,5.0,Minnesota,MRO,East North Central,-0.1,normal,2014-05-11,18:38:00,intentional attack,...,9.71,6.49,9.28,2640737,5457125,2.14,91.592666,NIGHT,False,Momentary
2,2010,10.0,Minnesota,MRO,East North Central,-1.5,cold,2010-10-26,20:00:00,severe weather,...,8.19,6.07,8.15,2586905,5310903,2.14,91.592666,NIGHT,False,Prolonged
3,2012,6.0,Minnesota,MRO,East North Central,-0.1,normal,2012-06-19,04:30:00,severe weather,...,9.25,6.71,9.19,2606813,5380443,2.14,91.592666,NIGHT,False,Prolonged
4,2015,7.0,Minnesota,MRO,East North Central,1.2,warm,2015-07-18,02:00:00,severe weather,...,10.16,7.74,10.43,2673531,5489594,2.14,91.592666,NIGHT,False,Prolonged


### UNIVARIATE ANALYSIS

In [358]:
px.histogram(outage, x='POPULATION')

In [359]:
px.histogram(outage, x='OUTAGE.DURATION')

In [360]:
climate_ct_group = outage.groupby('CLIMATE.CATEGORY')['OUTAGE.DURATION'].count()
px.bar(climate_ct_group,labels={'CLIMATE.CATEGORY':'Climate','value':'Count'})

In [361]:
px.histogram(outage, x='duration level')

In [367]:
duration_mean=outage[['YEAR', 'OUTAGE.DURATION']].groupby('YEAR').mean().reset_index()
px.line(duration_mean,x='YEAR', y='OUTAGE.DURATION')

### BIVARIATE ANALYSIS

In [ ]:
#fig 1
px.scatter(outage, x='POPULATION', y='OUTAGE.DURATION')

In [ ]:
#fig 4
px.histogram(outage, x='ANOMALY.LEVEL', color='CLIMATE.CATEGORY', barmode='overlay',
                    title='Conditional Distribution: ANOMALYLLYRL by CLIMATE CATEGORY',
                    labels={'ANOMALY.LEVEL': 'Anomaly Level', 'count': 'Frequency'})

In [ ]:
#6
pivot_table = pd.pivot_table(outage, values='OUTAGE.DURATION', index='CAUSE.CATEGORY', columns='CLIMATE.CATEGORY', aggfunc='count')
px.imshow(pivot_table, labels=dict(x="Climate Category", y="Cause Category", color="Outage Count"),
                 title='Outage Count by CAUSE CATEGORY and CLIMATE CATEGORY')

In [379]:
#7
state_max_anomaly = outage.groupby('U.S._STATE')['ANOMALY.LEVEL'].mean().reset_index()
px.bar(state_max_anomaly, x='U.S._STATE', y='ANOMALY.LEVEL',
              title='Maximum Anomaly Level by STATE',
              labels={'U.S._STATE': 'State', 'ANOMALY.LEVEL': 'Maximum Anomaly Level'})

In [ ]:
# fig1 = px.scatter(outage, x='POPULATION', y='OUTAGE.DURATION')
# fig2 = px.scatter(outage, x='ANOMALY.LEVEL', y='OUTAGE.DURATION',title='Scatter Plot: ANOMALYLLYRL vs. OUTAGE DURATION',
#                   labels={'ANOMALY.LEVEL': 'Anomaly Level', 'OUTAGE.DURATION': 'Outage Duration (hours)'})
# fig3 = px.box(outage, x='CAUSE.CATEGORY', y='OUTAGE.DURATION',
#               title='Box Plot: OUTAGE DURATION by CAUSE CATEGORY',
#               labels={'CAUSE.CATEGORY': 'Cause Category', 'OUTAGE.DURATION': 'Outage Duration (hours)'})
# fig4 = px.histogram(outage, x='ANOMALY.LEVEL', color='CLIMATE.CATEGORY', barmode='overlay',
#                     title='Conditional Distribution: ANOMALYLLYRL by CLIMATE CATEGORY',
#                     labels={'ANOMALY.LEVEL': 'Anomaly Level', 'count': 'Frequency'})
# climate_avg_duration = outage.groupby('CLIMATE.CATEGORY')['OUTAGE.DURATION'].mean().reset_index()
# fig5 = px.bar(climate_avg_duration, x='CLIMATE.CATEGORY', y='OUTAGE.DURATION',
#                title='Average Outage Duration by CLIMATE CATEGORY',
               
#                labels={'CLIMATE.CATEGORY': 'Climate Category', 'OUTAGE.DURATION': 'Average Outage Duration (hours)'})




# # Heatmap: Outage Count by CAUSE CATEGORY and CLIMATE CATEGORY
# pivot_table = pd.pivot_table(outage, values='OUTAGE.DURATION', index='CAUSE.CATEGORY', columns='CLIMATE.CATEGORY', aggfunc='count')
# fig6 = px.imshow(pivot_table, labels=dict(x="Climate Category", y="Cause Category", color="Outage Count"),
#                  title='Outage Count by CAUSE CATEGORY and CLIMATE CATEGORY')


# # Bar plot: Maximum Anomaly Level by STATE
# state_max_anomaly = outage.groupby('U.S._STATE')['ANOMALY.LEVEL'].mean().reset_index()
# fig7 = px.bar(state_max_anomaly, x='U.S._STATE', y='ANOMALY.LEVEL',
#               title='Maximum Anomaly Level by STATE',
#               labels={'U.S._STATE': 'State', 'ANOMALY.LEVEL': 'Maximum Anomaly Level'})
# fig8 = px.scatter(outage, x='ANOMALY.LEVEL', y='OUTAGE.DURATION')
# fig1.show()
# fig2.show()
# fig3.show()
# fig4.show()
# fig5.show()
# fig6.show()
# fig7.show()
# fig8.show()


### Interesting Aggregates

The pivot table tells the number of outage in different combination of NERC.REGION and CLIMATE.CATEGORY

In [386]:
pd.pivot_table(outage, values='YEAR', index='NERC.REGION', columns='CLIMATE.CATEGORY', aggfunc='count')

CLIMATE.CATEGORY,cold,normal,warm
NERC.REGION,,,
ECAR,2.0,22.0,8.0
FRCC,10.0,19.0,14.0
"FRCC, SERC",NaN,NaN,1.0
HECO,NaN,1.0,2.0
HI,1.0,NaN,NaN
MRO,13.0,25.0,6.0
NPCC,46.0,83.0,18.0
PR,NaN,1.0,NaN
RFC,149.0,219.0,48.0


The group table tells the mean of 'OUTAGE.DURATION', 'ANOMALY.LEVEL', and 'TOTAL.PRICE' in different states.

In [387]:
outage[['U.S._STATE', 'OUTAGE.DURATION', 'ANOMALY.LEVEL', 'TOTAL.PRICE']].groupby('U.S._STATE').mean()

,OUTAGE.DURATION,ANOMALY.LEVEL,TOTAL.PRICE
U.S._STATE,,,
Alabama,1152.800000,-0.060000,7.994000
Arizona,4552.920000,0.020000,9.282800
Arkansas,1514.360000,-0.104000,7.620800
California,1666.338384,-0.046970,13.060909
Colorado,901.071429,0.050000,9.218571
Connecticut,1278.833333,-0.244444,16.106667
Delaware,144.925000,-0.205000,11.140769
District of Columbia,4303.600000,-0.410000,12.274000
Florida,4094.666667,0.006667,9.155778


## Step 3: Assessment of Missingness

### NMAR

Among all columns with missing values, 'DEMAND.LOSS.MW' and 'CUSTOMERS.AFFECTED' has a high chance of being NMAR. Because these values could be missing due to the data collection method. As these values could bring direct and crucial negative effects to the power companies, they might not report the data to the institute collecting the data. 
With addition information about the companies such as their name, their size, or their annual revenue, it is possible to determine whether the missing values are dependent on the companies or their performances and thus making them MAR. 

### Missingness Dependency

We focus on the distribution of the column 'TOTAL.PRICE' and test it again the columns 'U.S._STATE' and 'CLIMATE.CATEGORY'.

In [315]:
def permutation_test(df, group_col, missing_col, n_permutations=1000):
    df=df.copy()
    df[missing_col]=df[missing_col].isna()
    group_proportions = df.groupby(group_col)[missing_col].mean()
    observed_tvd = 0.5*np.sum(np.abs(group_proportions.diff()))
    perm_tvds = []
    for _ in range(n_permutations):
        df_perm = df.copy()
        df_perm[missing_col] = np.random.permutation(df_perm[missing_col])
        perm_group_proportions = df_perm.groupby(group_col)[missing_col].mean()
        perm_tvd = 0.5 * np.sum(np.abs(perm_group_proportions.diff().fillna(0)))
        perm_tvds.append(perm_tvd)
    p_value = (np.array(perm_tvds) >= observed_tvd).mean()
    return {'p-val': p_value, 'obs': observed_tvd, 'result':perm_tvds}

In [316]:
def miss_graph(df, group_col, missing_col):
    df=df.copy()
    df[missing_col]=df[missing_col].isna()
    proportions = (
    df.groupby([group_col, missing_col]).size().reset_index(name="Count"))
    proportions["Proportion"] = proportions.groupby(group_col)["Count"]\
                                 .transform(lambda x: x / x.sum())
    return px.bar(proportions,x="Proportion",y=group_col,color=missing_col,\
                  orientation="h",barmode="group")

In [317]:
def miss_hits(result):
    return px.histogram(x=result['result'], histnorm="probability", title="Empirical Distribution of the TVD", \
                labels={"x": "TVD", "y": "Probability"})\
                    .add_vline(x=result['obs'], line_color="red",
                               annotation_text=f"Observed TVD = {result['obs']}",
                               annotation_position="top right")

1. 'U.S._STATE' vs.'TOTAL.PRICE' 
- Null Hypothesis: The distribution of 'U.S._STATE' is the same when 'TOTAL.PRICE' is missing vs not missing.
- Alternate Hypothesis: The distribution of 'U.S._STATE' is not the same when 'TOTAL.PRICE' is missing vs not missing.

In [318]:
first_result=permutation_test(outage, 'U.S._STATE', 'TOTAL.PRICE', n_permutations=1000)

In [319]:
miss_graph(outage, 'U.S._STATE', 'TOTAL.PRICE')

In [320]:
first_result['p-val']

np.float64(0.258)

In [321]:
first_result['obs']

np.float64(0.409892310205201)

In [322]:
miss_hits(first_result)

The observed TVD is around 0.4099 with the p-values of 0.256, which is higher than 0.05. As a result, we fail to reject the null hypothesis and conclude that the distribution of 'U.S._STATE' where 'TOTAL.PRICE' is missing has no significant difference from the distribution where 'TOTAL.PRICE' is not missing. Therefore, we can so far conclued that the missness of 'TOTAL.PRICE' does not depend on 'U.S._STATE'. 

2. 'CLIMATE.CATEGORY' vs. 'TOTAL.PRICE'
- Null Hypothesis: The distribution of 'CLIMATE.CATEGORY' is the same when 'TOTAL.PRICE' is missing vs not missing.
- Alternate Hypothesis: The distribution of ''CLIMATE.CATEGORY' is not the same when 'TOTAL.PRICE' is missing vs not missing.

In [323]:
second_result=permutation_test(outage, 'CLIMATE.CATEGORY', 'TOTAL.PRICE', n_permutations=1000)

In [324]:
miss_graph(outage, 'CLIMATE.CATEGORY', 'TOTAL.PRICE')

In [325]:
second_result['p-val']

np.float64(0.0)

In [326]:
second_result['obs']

np.float64(0.01643835616438356)

In [327]:
miss_hits(second_result)

The observed TVD is around 0.0164 with the p-values of 0.001, which is less than 0.05. As a result, we reject the null hypothesis and conclude that the distribution of 'CLIMATE.CATEGORY' where 'TOTAL.PRICE' is missing has a significant difference from the distribution where 'TOTAL.PRICE' is not missing. Therefore, we can say that the missing mechanism of 'TOTAL.PRICE' is MAR as it does depends on 'CLIMATE.CATEGORY'. 

## Step 4: Hypothesis Testing

We will be testing whether Duration is longer when it begins at night time. The columns that are use for testing are 'OUTAGE.DURATION' and 'DAY.OR.NIGHT'.

- Null Hypothesis: The average duration when it begins during night is the same as the duration that begins during day time. 
- Alternative Hypothesis: The average duration when it begins during night is longer than the duration that begins during day time.
- Test Statistic: Difference in mean (Mean duration begins during night - Mean duration begins during day time).

In [328]:
df=outage.copy()

level=0.01
obs=df[df['DAY.OR.NIGHT']=='NIGHT']['OUTAGE.DURATION'].mean()-\
    df[df['DAY.OR.NIGHT']=='DAY']['OUTAGE.DURATION'].mean()
print(f'observed stat: {obs}')
samp_diff=[]
n=10000
for i in range(n):
    df['DAY.OR.NIGHT']=np.random.permutation(df['DAY.OR.NIGHT'])
    diff=df[df['DAY.OR.NIGHT']=='NIGHT']['OUTAGE.DURATION'].mean()-\
        df[df['DAY.OR.NIGHT']=='DAY']['OUTAGE.DURATION'].mean()
    samp_diff.append(diff)

p=(samp_diff>obs).sum()/n
print(f'p-value: {p}')

if p<level:
    print('Reject null.')
    print('The average duration when it begins during night time is longer than '
    'the duration that begins during day time.')
else:
    print('Fail to reject null.')
    print('The average duration when it begins during night time is NOT '
    'longer than the duration that begins during day time.')

observed stat: 1085.1001534288862
p-value: 0.0
Reject null.
The average duration when it begins during night time is longer than the duration that begins during day time.


In [329]:
px.histogram(x=samp_diff, histnorm="probability", title="Distribution of Difference in Duration Mean", \
                labels={"x": "Difference (night-day)", "y": "Probability"})\
                    .add_vline(x=obs, line_color="red",
                               annotation_text=f"Observed Diff = {obs}",
                               annotation_position="top right")

The above graph show the observed difference vs. the differences' empirical distribution.

After performing a permutation test with 10000 simulations assuming that the null hypothesis was correct, we got the p-value of 0.0. Obviously, 0.0<0.05, implying a statistical significance, so we reject the null hypothesis and conclude that the average duration when it begins during night time is longer than the duration that begins during day time.

## Step 5: Framing a Prediction Problem

We want to classified the level of outage duration into 'Momentary', 'Short-term', 'Extended', 'Extended', and 'Chronic', so it would be a clssification problem. 
The metric we'll be using would be the acuracy because that's one the easist metric to be interpreted and we want the user to quickly understand our model's performance. 
During the prediction, user would have information to the following values: 'NERC.REGION', 'CLIMATE.REGION', 'CAUSE.CATEGORY', 'CAUSE.CATEGORY.DETAIL', 'DEMAND.LOSS.MW', 'CUSTOMERS.AFFECTED', 'ANOMALY.LEVEL', 'HAS.HURRICANE'. These information would be the input of our classifying prediction. 

## Step 6: Baseline Model

In [330]:
from sklearn.preprocessing import PolynomialFeatures, StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import mean_squared_error, r2_score, accuracy_score, f1_score
from sklearn.tree import DecisionTreeClassifier
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer, SimpleImputer, KNNImputer
from sklearn.ensemble import RandomForestClassifier
from sklearn.pipeline import Pipeline

In [331]:
X=outage[['NERC.REGION', 'CLIMATE.REGION', 'CAUSE.CATEGORY', 'CAUSE.CATEGORY.DETAIL', 'DEMAND.LOSS.MW', 'CUSTOMERS.AFFECTED', 'ANOMALY.LEVEL', 'HAS.HURRICANE']]
y=outage['duration level']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [332]:
#base line model
col_trans=ColumnTransformer(
    transformers=[
        ('one hot',OneHotEncoder(handle_unknown='ignore', drop='first'), ['NERC.REGION', 'CLIMATE.REGION'])
    ],
    remainder='drop'
)
base_mdl=Pipeline(steps=[
    ('trains', col_trans),
    ('dt', DecisionTreeClassifier(max_depth=10))
])
base_mdl.fit(X_train, y_train)

Pipeline(steps=[('trains',
                 ColumnTransformer(transformers=[('one hot',
                                                  OneHotEncoder(drop='first',
                                                                handle_unknown='ignore'),
                                                  ['NERC.REGION',
                                                   'CLIMATE.REGION'])])),
                ('dt', DecisionTreeClassifier(max_depth=10))])

In [333]:
#base line accuracy
y_pred = base_mdl.predict(X_test)
accuracy_score(y_test, y_pred)

0.35135135135135137

Our model is a DecisionTreeClassifier that uses two nominol features: 'NERC.REGION' and 'CLIMATE.REGION' to predict the duration level of a given outage. The information give by the model is useful for the govenment to determine whether this outage is serious or not.

We choose to use 'NERC.REGION' because it tells where the electricity came from in the region facing outage and choose 'CLIMATE.REGION' because it gives some information about the cause of outage as some sever whether exist in specific climate regions. Since both 'NERC.REGION' and 'CLIMATE.REGION' are categorical variables,we use OneHotEncoder to convert them to numeric values for training.

The model's performance in terms of accuracy is around 0.35 when tested on the test set, which is not a good value and means that the model requires critical inprovement during the next step. To do so, we plan to change the classify model, add more features for training, and use grid search to tune the hyperparameters.

## Step 7: Final Model

In [334]:
#final model pipelines
demand_trans=Pipeline(steps=[
    ('impute', IterativeImputer()),
    ('std', StandardScaler())
])
customer_trans=Pipeline(steps=[
    ('impute', KNNImputer(n_neighbors=5)),
    ('std', StandardScaler())
])
col_trans=ColumnTransformer(
    transformers=[
        ('demand impute', demand_trans, ['DEMAND.LOSS.MW']),
        ('customer impute', customer_trans, ['CUSTOMERS.AFFECTED']),
        ('one hot',OneHotEncoder(handle_unknown='ignore', drop='first'), ['NERC.REGION','CLIMATE.REGION', 'CAUSE.CATEGORY', 'CAUSE.CATEGORY.DETAIL', 'HAS.HURRICANE'])
    ],
    remainder='passthrough'
)
final_rf=Pipeline(steps=[
    ('trains', col_trans),
    ('forest', RandomForestClassifier(max_depth=10))
])

In [335]:
#final with grid
grid={
    'forest__n_estimators':[25, 50],
    'forest__max_depth':[10, 30, 50],
    'forest__min_samples_split':[20, 30]
}
rf_g_search=GridSearchCV(estimator=final_rf, param_grid=grid, cv=5, scoring='accuracy')
rf_g_search.fit(X_train, y_train)
rf_g_search.best_params_

/Users/the87/miniforge3/envs/dsc80/lib/python3.12/site-packages/sklearn/preprocessing/_encoders.py:242: UserWarning:

Found unknown categories in columns [0, 3] during transform. These unknown categories will be encoded as all zeros

/Users/the87/miniforge3/envs/dsc80/lib/python3.12/site-packages/sklearn/preprocessing/_encoders.py:242: UserWarning:

Found unknown categories in columns [3] during transform. These unknown categories will be encoded as all zeros

/Users/the87/miniforge3/envs/dsc80/lib/python3.12/site-packages/sklearn/preprocessing/_encoders.py:242: UserWarning:

Found unknown categories in columns [0, 3] during transform. These unknown categories will be encoded as all zeros

/Users/the87/miniforge3/envs/dsc80/lib/python3.12/site-packages/sklearn/preprocessing/_encoders.py:242: UserWarning:

Found unknown categories in columns [0, 3] during transform. These unknown categories will be encoded as all zeros

/Users/the87/miniforge3/envs/dsc80/lib/python3.12/site-packages/skl

{'forest__max_depth': 10,
 'forest__min_samples_split': 20,
 'forest__n_estimators': 50}

In [336]:
final_mdl=rf_g_search.best_estimator_
y_pred=final_mdl.predict(X_test)
accuracy_score(y_test, y_pred)

/Users/the87/miniforge3/envs/dsc80/lib/python3.12/site-packages/sklearn/preprocessing/_encoders.py:242: UserWarning:

Found unknown categories in columns [3] during transform. These unknown categories will be encoded as all zeros



0.5135135135135135

In our final model, we've decided to use a RandomForestClassifier to classify the duration level using the following features: 'NERC.REGION', 'CLIMATE.REGION', 'CAUSE.CATEGORY', 'CAUSE.CATEGORY.DETAIL', 'DEMAND.LOSS.MW', 'CUSTOMERS.AFFECTED', 'ANOMALY.LEVEL', 'HAS.HURRICANE'. 

Base on our baseline model, we added 6 new features. 
1. 'CAUSE.CATEGORY' (nominal): We add it because the way how outage happened can significantly impact its duration. We use OneHotEncoding to deal with it as it's a nominal feature.
2. 'CAUSE.CATEGORY.DETAIL' (nominal): We add it for the similar reason as 'CAUSE.CATEGORY', as it could help split when two levels have the same 'CAUSE.CATEGORY'. Similary, we use OneHotEncoding to deal with it because it's a nominal feature. 
3. 'DEMAND.LOSS.MW' (quantative): We added it because a outage with longer duration cause a larger loss in demand. We use IterativeImputer to impute the missing value because quite a lot of the values are missing but it provides important informations to the duration and we don't want to loss it. After imputing we standardized it to let the the values comparable. 
4. 'CUSTOMERS.AFFECTED' (quantative): We added it because ast he outage last longer, more people will be affected, so there'sa positive correlation between them. We use KNNImputer to impute the missing values because it does not have that much missing values compare to 'DEMAND.LOSS.MW' and we don't want to loss its important informations. 
5. 'ANOMALY.LEVEL' (quantative): We added 'ANOMALY.LEVEL' because it refleact to the climate status, specifically temperature, of the period of time and we believe temperature can affect people's electricity usage habits and further impact the outage duration. Since it's a quantative feature without missing values, we leave it as it was. 
6. 'HAS.HURRICANE' (nominal): We added it because hurricane causes serious damage, which can led to the postpone of outage restoration work. We use OneHotEncoding to deal with this nominal feature. 

We use GridSearchCV to find the best hyperparameters for our RandomForestClassifier.
The parameters are: 
- 'max_depth': 10
- 'min_samples_split': 20
- 'n_estimators': 50

Finally, our final model achieved the accuracy score of 0.54. Compare to the accuracy of 0.35 from the baseline model, our final model has a huge improvement on its accuracy, meaning that it's a way better model compare to the baseline model.

## Step 8: Fairness Analysis

Our groups for the fairness analysis are whether the outage happened with hurricane or not. We choose to use hurricane becuase it has a strong correlation between outage duration and we want to make sure that out model is predicing correctly. We use recall as our metric because we want to make sure that the model predicted correctly everytime there is a hurricane.

- Null Hypothesis: The model is fair. Its recall is the same whether there's hurricane or not, any difference should be due to random chance.
- Alternative Hypothesis: The model is unfair. Its recall is significantly different whether there's hurricane or not.

We use the recall_score of (hurricame - no hurricane) to record the result of the experiment 1000 times with the standard of 0.05.

In [337]:
from sklearn.metrics import recall_score

In [339]:
features=['NERC.REGION', 'CLIMATE.REGION', 'CAUSE.CATEGORY', 'CAUSE.CATEGORY.DETAIL',
           'DEMAND.LOSS.MW', 'CUSTOMERS.AFFECTED', 'ANOMALY.LEVEL', 'HAS.HURRICANE']

In [ ]:
has_hurr=outage[outage['HAS.HURRICANE']==True][features+['duration level']]
no_hurr=outage[outage['HAS.HURRICANE']==False][features+['duration level']]
true_has=has_hurr['duration level']
true_no=no_hurr['duration level']
def re_score(df):
    df=df.copy()
    df['HAS.HURRICANE']=np.random.permutation(df['HAS.HURRICANE'])
    has=df[df['HAS.HURRICANE']==True]
    no=df[df['HAS.HURRICANE']==False]
    pred_has=final_mdl.predict(has)
    pred_no=final_mdl.predict(no)
    return recall_score(true_has, pred_has, average='weighted')-recall_score(true_no, pred_no, average='weighted')
obs_has=final_mdl.predict(has_hurr[features])
obs_no=final_mdl.predict(no_hurr[features])
obs=recall_score(true_has, obs_has, average='weighted')-recall_score(true_no, obs_no, average='weighted')

test_result=[]
feature_df=outage[features]
for _ in range(1000):
    test_result.append(re_score(feature_df))
p=(np.array(test_result)>obs).sum()/1000

/Users/the87/miniforge3/envs/dsc80/lib/python3.12/site-packages/sklearn/preprocessing/_encoders.py:242: UserWarning:

Found unknown categories in columns [3] during transform. These unknown categories will be encoded as all zeros

/Users/the87/miniforge3/envs/dsc80/lib/python3.12/site-packages/sklearn/preprocessing/_encoders.py:242: UserWarning:

Found unknown categories in columns [3] during transform. These unknown categories will be encoded as all zeros

/Users/the87/miniforge3/envs/dsc80/lib/python3.12/site-packages/sklearn/preprocessing/_encoders.py:242: UserWarning:

Found unknown categories in columns [3] during transform. These unknown categories will be encoded as all zeros

/Users/the87/miniforge3/envs/dsc80/lib/python3.12/site-packages/sklearn/preprocessing/_encoders.py:242: UserWarning:

Found unknown categories in columns [3] during transform. These unknown categories will be encoded as all zeros

/Users/the87/miniforge3/envs/dsc80/lib/python3.12/site-packages/sklearn/prep

p-value: 0.0


/Users/the87/miniforge3/envs/dsc80/lib/python3.12/site-packages/sklearn/preprocessing/_encoders.py:242: UserWarning:

Found unknown categories in columns [3] during transform. These unknown categories will be encoded as all zeros

/Users/the87/miniforge3/envs/dsc80/lib/python3.12/site-packages/sklearn/preprocessing/_encoders.py:242: UserWarning:

Found unknown categories in columns [3] during transform. These unknown categories will be encoded as all zeros

/Users/the87/miniforge3/envs/dsc80/lib/python3.12/site-packages/sklearn/preprocessing/_encoders.py:242: UserWarning:

Found unknown categories in columns [3] during transform. These unknown categories will be encoded as all zeros

/Users/the87/miniforge3/envs/dsc80/lib/python3.12/site-packages/sklearn/preprocessing/_encoders.py:242: UserWarning:

Found unknown categories in columns [3] during transform. These unknown categories will be encoded as all zeros



In [353]:
f'p-value:{p}'

'p-value:0.0'

In [354]:
px.histogram(x=test_result, histnorm="probability", title="Distribution of Recall Difference for Has or Doesn't have Hurricane", \
                labels={"x": "Recall Difference (has-no)", "y": "Probability"})\
                    .add_vline(x=obs, line_color="red",
                               annotation_text=f"Observed Diff = {obs}",
                               annotation_position="top right")

We get the p-value of 0.0 after the test, which is less than the standard 0.05. Therefore, we reject the null hypothesis and conclude that our model is unfair as the significant different in recall from the two groups. 

References:
- https://docs.lib.purdue.edu/cgi/viewcontent.cgi?params=/context/civeng/article/1035/&path_info=10.1016_j.dib.2018.06.067.pdf